### PR TITLE
`LocationTag.points_around_x/y/z` tags

### PR DESCRIPTION
**Pull request by acikek**
## Additions

- `LocationTag.points_around_x/y/z`

### Notes

- I put the logic for the `rotate_around` tags in separate methods since the `points_around` tags now share them.

image
*Look at this. It's cool, right?*

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java [hunk 1]
```diff
@@ -2291,17 +2307,133 @@ else if (yaw < 315) {
...
+            }
+            if (amount.asInt() < 1) {
+                attribute.echoError("Invalid amount of points! There must be at least 1 point.");
+            }
+            double angle = 2 * Math.PI / amount.asInt();
+            ListTag points = new ListTag();
+            for (int i = 1; i <= amount.asInt(); i++) {
+                double[] result = getRotatedAroundX(angle * i, radius.asDouble(), 0);
+                LocationTag newLocation = object.clone();
+                newLocation.add(0, result[0], result[1]);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java]: Combine the above two lines into one... never type `somevec.add(x,y,z);` on its own line (or any similar method) as that, uh... technically makes sense to do under Bukkit's vector API but doesn't make sense in any others, and trusting Bukkit's semantics is a bad habit to build

File: plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java [hunk 2]
```diff
@@ -2291,17 +2307,133 @@ else if (yaw < 315) {
...
+            MapTag inputMap = attribute.paramAsType(MapTag.class);
+            if (inputMap == null) {
+                return null;
+            }
+            ObjectTag radiusObj = inputMap.getObject("radius");
+            ObjectTag pointsObj = inputMap.getObject("points");
+            if (radiusObj == null || pointsObj == null) {
+                return null;
+            }
+            ElementTag radius = radiusObj.asType(ElementTag.class, attribute.context);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java]: `asElement()`

File: plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java [hunk 3]
```diff
@@ -2291,17 +2307,133 @@ else if (yaw < 315) {
...
+            ElementTag amount = pointsObj.asType(ElementTag.class, attribute.context);
+            if (radius == null || amount == null) {
+                return null;
+            }
+            if (amount.asInt() < 1) {
+                attribute.echoError("Invalid amount of points! There must be at least 1 point.");
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java]: `return null;` after the error

File: plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java [hunk 4]
```diff
@@ -2291,17 +2307,133 @@ else if (yaw < 315) {
...
+            }
+            double angle = 2 * Math.PI / amount.asInt();
+            ListTag points = new ListTag();
+            for (int i = 1; i <= amount.asInt(); i++) {
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java]: You called `asInt` 3 times, the last of which is inside the body of a loop - this means it'll parse from input text to an int... however many times the `amount` is plus 2 more. Store the integer in a variable.

File: plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java [hunk 5]
```diff
@@ -2291,17 +2307,133 @@ else if (yaw < 315) {
...
+        // @group math
+        // @description
+        // Returns a list of points in a circular rotation around a location's y axis with the specified radius and number of points.
+        // For example: <player.location.points_around_y[radius=10;points=16]>
+        // -->
+        tagProcessor.registerTag(ListTag.class, "points_around_y", (attribute, object) -> {
+            if (!attribute.hasParam()) {
+                return null;
+            }
+            MapTag inputMap = attribute.paramAsType(MapTag.class);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/objects/LocationTag.java]: The body of the tags could probably be simplified down to one single method, that each of the three tags, considering 99% of the body is exactly the same.

**Comment by acikek** [comment]
Addressed the issues, condensed the parsing into a method that they all call now. Tested and verified to be working still.

**Pull request merged**